### ISSUE TITLE
How to check if a certain entity still exists?

### ISSUE THREAD

**Issue by Anatoly03** [NateTheGreatt/bitECS#68]
During a bug in my own code I noticed that the `Comp.attr[id]` notation will also give a value, even if the entity with `id` is already removed. In my implementation, I had a `ForceArrow` component which takes two other entities and applies velocity modifications to them.

It would be nice to have something like `exists(world, eid)` if possible.

**Comment by NateTheGreatt** [comment]
an explicit API method might be useful. however, one way to do this without an addition to the API would be to define a query like so:

```js
const getAllEntities = defineQuery()

const allEnts = getAllEntities(world)

allEnts.includes(eid)
```

this is what the new API method would do, anyways. so there are two options here:

1. add new API method
2. add documentation for the above

what do you think?

**Comment by NateTheGreatt** [comment]
on second thought, i can speed up the lookup time if i make an API method for it, because underneath that array is a sparse set.

i'll add a new API for this on next release 🙂 thanks for creating the issue!

**Comment by Anatoly03** [comment]
@NateTheGreatt Thank you very much! Appreciated!

**Comment by NateTheGreatt** [comment]
@Anatoly03 i realized there is a caveat here that will require me to introduce the concept of entity generations. because `bitECS` recycles entity IDs, it is possible a new entity will be created with the same ID of a removed entity, causing a false-positive return value from a method like `entityExists` and even from the above `getAllEntities` example.

the solution is to store a generation ID in the upper bits of an entityID much like `flecs` does. i will be implementing this as soon as i can!

**Comment by Anatoly03** [comment]
> @Anatoly03 i realized there is a caveat here that will require me to introduce the concept of "entity generations".

Oh only if it’s easy to implement! I don’t want you to take all the effort for one simple feature!

**Comment by speigg** [comment]
@NateTheGreatt being able to track whether or not an entity exists without mistaking it for a previously created and destroyed entity is an important feature! Entity generation ID sounds like a decent solution :)

**Comment by NateTheGreatt** [comment]
@speigg entity generations is the ideal solution, and am working on implementing that along with archetypes. however, i've decided as an immediate interim solution to simply prevent EID recycling until a certain number of entities have been removed, and then treating the list of removed entities as a FIFO queue.

**Comment by NateTheGreatt** [comment]
@Anatoly03 @speigg @yairhaimo @vegeta897

i have released `bitecs@0.3.35` which includes `entityExists` along with a delayed entity ID recycling to help prevent false-positive existence checks :+1:
